go-controller/pkg/node/node_smartnic.go

```diff
@@ -0,0 +1,203 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/cni"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/kube"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+//watchSmartNicPods watch updates for pod smart nic annotations
+func (n *OvnNode) watchSmartNicPods() {
+	var retryPods sync.Map
+	// servedPods tracks the pods that got a VF
+	var servedPods sync.Map
+	_ = n.watchFactory.AddPodHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*kapi.Pod)
+			klog.Infof("AddFunc for POD: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
+			if !util.PodWantsNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
+				return
+			}
+			if util.PodScheduled(pod) {
+				// Is this pod created on same node where the smart NIC
+				if n.name != pod.Spec.NodeName {
+					return
+				}
+
+				vfRepName, err := n.getVfRepName(pod)
+				if err != nil {
+					klog.Infof("Failed to get rep name, %s. retrying", err)
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				podInterfaceInfo, err := cni.PodAnnotation2PodInfo(pod.Annotations, true)
+				if err != nil {
+					retryPods.Store(pod.UID, true)
+					return
+				}
+				err = n.addRepPort(pod, vfRepName, podInterfaceInfo)
+				if err != nil {
+					klog.Infof("Failed to add rep port, %s. retrying", err)
+					retryPods.Store(pod.UID, true)
+				} else {
+					servedPods.Store(pod.UID, true)
+				}
+			} else {
+				// Handle unscheduled pods later in UpdateFunc
+				retryPods.Store(pod.UID, true)
+				return
+			}
+		},
+		UpdateFunc: func(old, newer interface{}) {
+			pod := newer.(*kapi.Pod)
+			klog.Infof("UpdateFunc for POD: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
+			if !util.PodWantsNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
+				retryPods.Delete(pod.UID)
+				return
+			}
+			_, retry := retryPods.Load(pod.UID)
+			if util.PodScheduled(pod) && retry {
+				if n.name != pod.Spec.NodeName {
+					retryPods.Delete(pod.UID)
+					return
+				}
+				vfRepName, err := n.getVfRepName(pod)
+				if err != nil {
+					klog.Infof("Failed to get rep name, %s. retrying", err)
+					retryPods.Store(pod.UID, true)
```

Comment:
Yes, this is needed just for the first time.
Looking at the implementation of `Store()`, it will re-assign the value instead of skip as it operates on pointers.

Reply:
removed my original reply here,

looking at this agian, i believe you are right as L#69 ensures that

Reply:
but it's still in this commit